src/classes/CCON_Faculty_TDTM.cls

```diff
@@ -96,29 +95,45 @@ public with sharing class CCON_Faculty_TDTM extends TDTM_Runnable {
                     for (Integer i = 0; i < newlist.size(); i++) {
                         Course_Enrollment__c enroll = (Course_Enrollment__c)newlist[i];
 
-                        if(enroll.get('RecordTypeId') == facultyRecType) {
-                            enroll.Account__c = offerings.get(enroll.Course_Offering__c).Course__r.Account__c;
-
-                            if (contactIdAccountIdAfflList.containsKey(enroll.Contact__c) && (
-                                 contactIdAccountIdAfflList.get(enroll.Contact__c).containsKey(enroll.Account__c))) {
-                                enroll.Affiliation__c = contactIdAccountIdAfflList.get(enroll.Contact__c).get(enroll.Account__c)[0].Id;
-                            } else {
-                                Affiliation__c affl = new Affiliation__c(
-                                    Contact__c = enroll.Contact__c,
-                                    Account__c = enroll.Account__c,
-                                    Role__c = 'Faculty'
-                                );
-                                affiliationsToInsert.add(affl);
-                                connsNeedingAffls.add(enroll);
+                        // Skip over the Course Connection if it is not for a Faculty.
+                        if(enroll.get('RecordTypeId') != facultyRecType) {
+                            continue;
+                        }
+
+                        enroll.Account__c = offerings.get(enroll.Course_Offering__c).Course__r.Account__c;
+
+                        Map<Id, List<Affiliation__c>> affiliationsByAccountId = contactIdAccountIdAfflList.containsKey(enroll.Contact__c) ? 
+                            contactIdAccountIdAfflList.get(enroll.Contact__c) : null;
+
+                        if (affiliationsByAccountId != null && affiliationsByAccountId.containsKey(enroll.Account__c)) {
+                            enroll.Affiliation__c = affiliationsByAccountId.get(enroll.Account__c)[0].Id;
+                        } else {
+                            connsNeedingAffls.add(enroll);
+                            
+                            // Is the Affiliation already being inserted? If so, do not insert a duplicate.
+                            String contactAccountKey = enroll.Contact__c + '' + enroll.Account__c;
+                            if (affiliationsToInsertByContactAccountString.containsKey(contactAccountKey)) {
+                                continue;
                             }
+
+                            Affiliation__c affl = new Affiliation__c(
```

Comment:
Note that here we don't set the Start & End Date of the Affiliation, but we do in the COFF_Affiliation_TDTM class when Course Connections is disabled. I noted in the WI that we should consider making this consistent (but out of scope of this bug fix)